### PR TITLE
Update GEMM

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

* Organize and collect metal subroutine templates and elements in `metal/kernels/steel/`
* Update gemm elements for better performance 
* Add split-K specialization for gemm
* Add `addmm` primitive, op and bindings for fused matmul and bias addition 
* Update tests and benchmarks as needed 

I'm attaching a collection of the benchmarks I ran in a txt below
[bench.txt](https://github.com/ml-explore/mlx/files/13894895/bench.txt)

## Checklist

Put an `x` in the boxes that apply.

- [`x` ] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [`x` ] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [`x` ] I have added tests that prove my fix is effective or that my feature works
- [ `x`] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
🤤

**Comment by awni** [comment]
@jagrit06 I will review very soon, can't wait to get this in. In the meantime, can you:

1. Rebase and resolve the conflict:
2. I get a compiler warning:

```
/Users/awni/mlx/mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm_addmm.metal:195:16: warning: unknown attribute 'unlikely' ignored
```

Could you check that and clear it if possible?

File: mlx/ops.cpp [hunk 1]
```diff
@@ -3057,4 +3057,110 @@ array inner(const array& a, const array& b, StreamOrDevice s /* = {} */) {
...
+
+  if (a.dtype() != out_type) {
+    a = astype(a, out_type, s);
+  }
+  if (b.dtype() != out_type) {
+    b = astype(b, out_type, s);
+  }
+  if (c.dtype() != out_type) {
+    c = astype(c, out_type, s);
+  }
```
Comment (angeloskath) [inline mlx/ops.cpp:3116]: The `if` is unnecessary since `astype` does exactly that.

File: mlx/ops.cpp [hunk 2]
```diff
@@ -3057,4 +3057,110 @@ array inner(const array& a, const array& b, StreamOrDevice s /* = {} */) {
...
+  if (a.shape(-1) != b.shape(-2)) {
+    std::ostringstream msg;
+    msg << "[addmm] Last dimension of first input with shape " << a.shape()
+        << " must match second to last dimension of"
+        << " second input with shape " << b.shape() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // Type promotion
+  auto out_type = promote_types(a.dtype(), b.dtype());
```
Comment (angeloskath) [inline mlx/ops.cpp:3099]: Maybe `result_type({a, b, c})`?

File: mlx/backend/common/default_primitives.cpp [hunk 3]
```diff
@@ -148,4 +148,58 @@ void Matmul::eval_cpu(const std::vector<array>& inputs, array& out) {
...
+        a.data<float>() + elem_to_loc(M * K * i, a.shape(), a.strides()),
+        lda,
+        b.data<float>() + elem_to_loc(K * N * i, b.shape(), b.strides()),
+        ldb,
+        beta_, // beta
+        out.data<float>() + M * N * i,
+        out.shape(-1) // ldc
+    );
+  }
+}
```
Comment (awni) [inline mlx/backend/common/default_primitives.cpp:203]: It seems like we should refactor this into a common function used by the one above, it's almost the same except for alpha and beta right?

**Review by angeloskath — APPROVED** [review]
It looks absolutely great! Even more so given the perf improvements!

I left a small number of questions.

File: mlx/ops.cpp [hunk 4]
```diff
@@ -3057,4 +3057,110 @@ array inner(const array& a, const array& b, StreamOrDevice s /* = {} */) {
   return tensordot(a, b, {{-1}, {-1}}, s);
 }
 
+/** Compute the matrix product of A and B and add C to the result */
+array addmm(
+    const array& in_c,
+    const array& in_a,
+    const array& in_b,
```
Comment (awni) [inline mlx/ops.cpp:3064]: I would just make this pass by value instead. Since we always copy them anyway. I think it would be good to start moving to pass by value for these cases as it gives us the opportunity to use move semantics

File: python/src/ops.cpp [hunk 5]
```diff
@@ -3476,4 +3476,34 @@ void init_ops(py::module_& m) {
...
+      "b"_a,
+      py::pos_only(),
+      "alpha"_a = 1.0f,
+      "beta"_a = 1.0f,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+        addmm(c: array, a: array, b: array, /, alpha: float = 1.0, beta: float = 1.0,  *, stream: Union[None, Stream, Device] = None) -> array
+
+        Matrix multiplication and linear combination.
```
Comment (awni) [inline python/src/ops.cpp:3493]: I think `Matrix multiplication with addition.` is more clear.

File: python/src/ops.cpp [hunk 6]
```diff
@@ -3476,4 +3476,34 @@ void init_ops(py::module_& m) {
...
+        Perform the (possibly batched) matrix multiplication of two arrays and add to the result
+        with optional scaling factors.
+
+        Args:
+            c (array): Input array or scalar.
+            a (array): Input array or scalar.
+            b (array): Input array or scalar.
+            alpha (float, optional): Scaling factor for the 
+                matrix product of ``a`` and ``b`` (default: 1)
+            beta (float, optional): Scaling factor for ``c`` (default: 1)
```
Comment (awni) [inline python/src/ops.cpp:3504]: Minor nit: use double backticks on the default values

File: python/src/ops.cpp [hunk 7]
```diff
@@ -3476,4 +3476,34 @@ void init_ops(py::module_& m) {
...
+
+        Returns:
+            array: ``alpha`` * (``a`` @ ``b``)  + ``beta`` * ``c``
```
Comment (awni) [inline python/src/ops.cpp:3507]:
```suggestion
            array: ``alpha * (a @ b)  + beta * c``
```

Is there a reason you inlined coded only the variables? Does it look weird o/w?

File: python/tests/test_blas.py [hunk 8]
```diff
@@ -444,3 +445,139 @@ def test_matrix_vector_edgecases(self):
...
+            c_npy = np.ones(c_shape).astype(np.float32)
+            c_mlx = mx.array(c_npy)
+
+            d_npy = alpha * (a_npy @ b_npy) + beta * c_npy
+            d_mlx = mx.addmm(c_mlx, a_mlx, b_mlx, alpha, beta)
+
+            self.assertListEqual(list(d_npy.shape), list(d_mlx.shape))
+            self.assertTrue(np.allclose(d_mlx, d_npy, atol=1e-5))
+
+    def test_addmm_grad(self):
```
Comment (awni) [inline python/tests/test_blas.py:542]: 👍

**Review by awni — COMMENTED** [review]
@jagrit06 ubelievable!

Left a few cosmetic suggestions. Could you do one more thing and make sure:

I think it would be good to run one or two "integration" style tests. E.g. double check MNIST and Transformer training still works.

Other than that + comments, this should be good to go.

File: mlx/backend/metal/kernels/steel/gemm/transforms.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    return static_cast<OutT>(x * alpha + (beta * c));
+  }
+};
+
+template <typename T>
+struct AccumHelper {
+  typedef float accum_type;
+};
+
+struct BlockSwizzle {
```
[the remark this answers is not in the document]
Comment (jagrit06) [reply mlx/backend/metal/kernels/steel/gemm/transforms.h:52]: Honestly, just as a matter of convenience till I sort out the API and organization a bit more

File: mlx/backend/common/default_primitives.cpp (hunk 3 above)
Comment (jagrit06) [reply mlx/backend/common/default_primitives.cpp:203]: Good point, I did exactly did that for the accelerate backend, I'll do the same here

File: mlx/ops.cpp (hunk 4 above)
Comment (jagrit06) [reply mlx/ops.cpp:3064]: Got it 
Should I update the `matmul` op as well ?

File: python/src/ops.cpp (hunk 5 above)
Comment (jagrit06) [reply python/src/ops.cpp:3493]: Should I make it `Matrix multiplication with addition and optional scaling` or am I overthinking ?

File: mlx/backend/metal/matmul.cpp [hunk 10]
```diff
@@ -209,11 +229,108 @@ void mlx_matmul(
...
+
+    compute_encoder->setBytes(&params, sizeof(GEMMSpiltKParams), 3);
+    compute_encoder->dispatchThreadgroups(grid_dims, group_dims);
+
+    // Do accum kernel
+    {
+      auto c_split_buf =
+          static_cast<const MTL::Resource*>(C_split.buffer().ptr());
+      const class MTL::Resource* const resources[1] = {c_split_buf};
+      compute_encoder->memoryBarrier(resources, 1);
```
[the remark this answers is not in the document]
Comment (jagrit06) [reply mlx/backend/metal/matmul.cpp:300]: Good catch! This is a holdover from some debugging I was doing, it isn't needed at the moment

File: python/src/ops.cpp (hunk 5 above)
Comment (awni) [reply python/src/ops.cpp:3493]: Yea! Might as well!

File: mlx/ops.cpp (hunk 4 above)
Comment (jagrit06) [reply mlx/ops.cpp:3064]: Scratch that, we need the originals for padding dims in gemv cases there

**Review by awni — APPROVED** [review]
🚀 merge when ready.

**Comment by awni** [comment]
@altaic good point. I wasn't planning to include benchmarks in the release notes themselves as there is quite a bit of variability between performance changes based on machine and size and it's difficult to do something lightweight without being misleading and hard to do something accurate without being too dense 😄 . But I will give it some thought, would be nice to be a bit more concrete.

**Comment by awni** [comment]
Adding some benchmarks for future reference and to link to in release notes:

These were recorded on an M2 Ultra

### Micro benchmarks

|Batch|M   |N   |K    |Type   |Transpose flags|Pre FLOPS|Post FLOPS|Improvement|
|-----|----|----|-----|-------|---------------|---------|----------|-----------|
|16   |234 |768 |3072 |float32|nn             |12027.758|13231.401 |10.01%     |
|1    |64  |64  |25344|float32|nn             |467.469  |1376.638  |194.49%    |
|16   |1024|1024|1024 |float32|nn             |19055.476|19013.485 |-0.22%     |
|1    |1024|1024|2048 |float32|nn             |13469.525|13122.696 |-2.57%     |
|4    |1024|1024|4096 |float32|nn             |17611.786|17418.788 |-1.10%     |
|4    |1024|4096|1024 |float32|nn             |19113.772|19288.433 |0.91%      |
|1    |4096|4096|4096 |float32|nn             |19797.157|20464.456 |3.37%      |
|16   |234 |768 |3072 |float32|nt             |11754.559|11603.867 |-1.28%     |
|1    |64  |64  |25344|float32|nt             |433.129  |1650.841  |281.14%    |
|16   |1024|1024|1024 |float32|nt             |16364.466|18041.09  |10.25%     |
|1    |1024|1024|2048 |float32|nt             |12519.554|12233.901 |-2.28%     |
|4    |1024|1024|4096 |float32|nt             |15732.043|16663.428 |5.92%      |
|4    |1024|4096|1024 |float32|nt             |16561.408|18245.227 |10.17%     |
|1    |4096|4096|4096 |float32|nt             |17262.061|19372.88  |12.23%     |
|16   |234 |768 |3072 |float32|tn             |12060.507|12076.578 |0.13%      |
|1    |64  |64  |25344|float32|tn             |460.053  |1601.016  |248.01%    |
|16   |1024|1024|1024 |float32|tn             |18558.581|18893.906 |1.81%      |
|1    |1024|1024|2048 |float32|tn             |13341.663|12418.367 |-6.92%     |
|4    |1024|1024|4096 |float32|tn             |18062.5  |17356.457 |-3.91%     |
|4    |1024|4096|1024 |float32|tn             |18602.391|19119.891 |2.78%      |
|1    |4096|4096|4096 |float32|tn             |19456.247|20389.987 |4.80%      |
|16   |234 |768 |3072 |float16|nn             |15269.618|15311.032 |0.27%      |
|1    |64  |64  |25344|float16|nn             |509.349  |1724.787  |238.63%    |
|16   |1024|1024|1024 |float16|nn             |20864.6  |21238.192 |1.79%      |
|1    |1024|1024|2048 |float16|nn             |14822.563|13736.598 |-7.33%     |
|4    |1024|1024|4096 |float16|nn             |20463.982|19009.315 |-7.11%     |
|4    |1024|4096|1024 |float16|nn             |20845.136|21281.805 |2.09%      |
|1    |4096|4096|4096 |float16|nn             |21655.472|22271.052 |2.84%      |
|16   |234 |768 |3072 |float16|nt             |15003.988|15817.855 |5.42%      |
|1    |64  |64  |25344|float16|nt             |569.462  |1712.874  |200.79%    |
|16   |1024|1024|1024 |float16|nt             |19566.767|21604.038 |10.41%     |
|1    |1024|1024|2048 |float16|nt             |13343.281|14436.931 |8.20%      |
|4    |1024|1024|4096 |float16|nt             |19267.185|20849.905 |8.21%      |
|4    |1024|4096|1024 |float16|nt             |19626.402|21608.829 |10.10%     |
|1    |4096|4096|4096 |float16|nt             |20224.786|22677.265 |12.13%     |
|16   |234 |768 |3072 |float16|tn             |14778.14 |14845.274 |0.45%      |
|1    |64  |64  |25344|float16|tn             |509.216  |1677.493  |229.43%    |
|16   |1024|1024|1024 |float16|tn             |20794.861|21289.926 |2.38%      |
|1    |1024|1024|2048 |float16|tn             |14485.417|13602.136 |-6.10%     |
|4    |1024|1024|4096 |float16|tn             |20126.936|18871.253 |-6.24%     |
|4    |1024|4096|1024 |float16|tn             |20778.222|21338.947 |2.70%      |
|1    |4096|4096|4096 |float16|tn             |21362.745|22405.331 |4.88%      |

**Comment by awni** [comment]
And more end-to-end benchmarks also recorded on an M2 Ultra:

### Transformer training

From `mlx-examples/transformer_lm` do `python main.py --gpu`:

Pre: It/sec 5.043
Post: It/sec 5.204

### LoRA fine-tuning

From `mlx-examples/lora` do
```
python lora.py --model mistralai/Mistral-7B-v0.1 --train --steps-per_report 40
```

Pre: 446 tokens per second
Post: 462 tokens per second


Note I also validated the loss looks good for both.
